src/Interpreters/GraceHashJoin.cpp

```diff
@@ -526,17 +559,11 @@ IBlocksStreamPtr GraceHashJoin::getDelayedBlocks()
 
     if (hash_join)
     {
-        auto right_blocks = hash_join->releaseJoinedBlocks();
-        Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, right_blocks, buckets.size());
-
-        for (size_t i = 0; i < blocks.size(); ++i)
+        auto right_blocks = hash_join->releaseJoinedBlocks(/* restructure */ false);
+        for (auto & block : right_blocks)
         {
-            if (blocks[i].rows() == 0 || i == bucket_idx)
-                continue;
-
-            if (i < bucket_idx)
-                throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected bucket index {} when current bucket is {}", i, bucket_idx);
-            buckets[i]->addRightBlock(blocks[i]);
+            Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, block, buckets.size());
+            flushBlocksToBuckets<JoinTableSide::Right>(blocks, buckets, bucket_idx);
```

Comment:
Scatter and flush blocks one by one. Otherwise we did at least 2x allocation in `Blocks blocks = JoinCommon::scatterBlockByHash(right_key_names, right_blocks, buckets.size());`